contracts/multisig-prover/src/encoding/mvx/mod.rs

```diff
@@ -0,0 +1,438 @@
+use bech32::FromBase32;
+use cosmwasm_std::Uint256;
+use multiversx_sc_codec::top_encode_to_vec_u8;
+use multiversx_sc_codec::{EncodeErrorHandler, NestedEncode, NestedEncodeOutput};
+use sha3::{Digest, Keccak256};
+
+use axelar_wasm_std::hash::Hash;
+use multisig::key::PublicKey;
+use multisig::msg::Signer;
+use multisig::verifier_set::VerifierSet;
+use router_api::Message as RouterMessage;
+
+use crate::error::ContractError;
+use crate::payload::Payload;
+
+pub mod execute_data;
+
+#[derive(Debug)]
+pub struct Message {
+    pub source_chain: String,
+    pub message_id: String,
+    pub source_address: String,
+    pub contract_address: [u8; 32],
+    pub payload_hash: [u8; 32],
+}
+
+#[derive(PartialEq)]
+pub struct WeightedSigner {
+    pub signer: [u8; 32],
+    pub weight: Vec<u8>,
+}
+
+#[derive(PartialEq)]
+pub struct WeightedSigners {
+    pub signers: Vec<WeightedSigner>,
+    pub threshold: Vec<u8>,
+    pub nonce: [u8; 32],
+}
+
+impl NestedEncode for Message {
+    fn dep_encode_or_handle_err<O, H>(&self, dest: &mut O, h: H) -> Result<(), H::HandledErr>
+    where
+        O: NestedEncodeOutput,
+        H: EncodeErrorHandler,
+    {
+        self.source_chain.dep_encode_or_handle_err(dest, h)?;
+        self.message_id.dep_encode_or_handle_err(dest, h)?;
+        self.source_address.dep_encode_or_handle_err(dest, h)?;
+        self.contract_address.dep_encode_or_handle_err(dest, h)?;
+
+        self.payload_hash.dep_encode_or_handle_err(dest, h)
+    }
+}
+
+impl NestedEncode for WeightedSigner {
+    fn dep_encode_or_handle_err<O, H>(&self, dest: &mut O, h: H) -> Result<(), H::HandledErr>
+    where
+        O: NestedEncodeOutput,
+        H: EncodeErrorHandler,
+    {
+        self.signer.dep_encode_or_handle_err(dest, h)?;
+
+        self.weight.dep_encode_or_handle_err(dest, h)
+    }
+}
+
+impl NestedEncode for WeightedSigners {
+    fn dep_encode_or_handle_err<O, H>(&self, dest: &mut O, h: H) -> Result<(), H::HandledErr>
+    where
+        O: NestedEncodeOutput,
+        H: EncodeErrorHandler,
+    {
+        self.signers.dep_encode_or_handle_err(dest, h)?;
+        self.threshold.dep_encode_or_handle_err(dest, h)?;
+
+        self.nonce.dep_encode_or_handle_err(dest, h)
+    }
+}
+
+impl WeightedSigners {
+    pub fn hash(&self) -> Hash {
+        let mut encoded = Vec::new();
+
+        for signer in self.signers.iter() {
+            encoded.push(signer.signer.as_slice());
+            encoded.push(signer.weight.as_slice());
+        }
+
+        encoded.push(self.threshold.as_slice());
+        encoded.push(self.nonce.as_slice());

```

Comment:
Good question. Because the result from the below `encode` function is not the same as this result.
The Mvx encoding used there will encode the signers using NestedEncode (which is implemented for `WeightedSigners` and `WeightedSigner`), and the result is not the same as here in the `hash` function which uses top encode.

In MultiversX encoding there is a difference between top and nested encoding, where top encoding is more efficient, but nested encoding needs to be used sometimes.


In the contract the hash is also calculated like this because it is more efficient, hence the differences.